design/design-doc-integration-tests.md

```diff
@@ -0,0 +1,333 @@
+# Crossplane Integration Testing Framework
+* Owner: Dan Mangum (@hasheddan)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+As Crossplane and its stacks have grown and evolved, the surface area for
+potential bugs has increased as well. Every new controller in a stack has the
+potential to behave in an undesirable manner given any number of edge-cases.
+Currently, automated integration tests are performed at the stack level via
+[bash scripts][current-integration-tests] that are triggered in a Jenkins
+pipeline. While this is effective for determining if the stack can successfully
+be installed into a Crossplane Kubernetes cluster, it does not go so far as to
+test any of the controllers beyond that they start.
+
+All other testing is performed in a manual ad-hoc manner for individual PR's or
+leading up to a release. This leads to uncertainty around how last-minute
+changes landing at the same time will affect the functionality of stacks.
+
+## Goals
```

Comment:
Are you thinking that these integration tests would hit real cloud provider APIs, or use fake ones mocked out with something like https://golang.org/pkg/net/http/httptest/ or https://github.com/dnaeon/go-vcr? My suspicion is there would be value in supporting both.

I think we'd get value on top of our existing unit tests by testing a controller as a whole even if we mocked out the cloud provider APIs, because we'd be exercising how the controller as a whole functions against a real API server when you manipulate resources. You could write, for example, a test that submits a `CloudSQLInstanceClass`, submits a `MySQLInstance`, and ensures everything works as expected with a fake CloudSQL API.

Using the _real_ CloudSQL API would further increase confidence, but costs more money, is more fragile, and would need us to support injecting cloud provider credentials in a secure fashion.

Reply:
Hmm yes it was definitely my intention to run against the real provider APIs. I agree that the mocks would be a slight improvement over the unit tests, but anecdotally it wouldn't really increase my confidence of the controllers.

> Using the real CloudSQL API would further increase confidence, but costs more money, is more fragile, and would need us to support injecting cloud provider credentials in a secure fashion.

* "costs more money": this is definitely true when compared to testing against mock APIs. However, I would argue that if we are running a bunch of manual tests on a regular basis, automating those and running them periodically (maybe a few times per day), could actually be more cost-efficient or at least only a marginal increase.

* "is more fragile": Crossplane is run against the real APIs when it is being used. I think running these tests in a controlled environment is not only beneficial, but necessary.

* "injecting cloud provider credentials": because cloud provider credentials are injected into Crossplane in a consistent manner (i.e. as a k8s secret), I believe this will be easier than anticipated. A hook to do this is definitely a common area that could be moved into the framework itself (maybe there is a `Setup` component in addition to the `Clean`).

Reply:
My suspicion is that even using mocks would be more than a slight improvement on the cloud provider APIs, especially for integration testing the full dynamic and static provisioning scenarios rather than just managed resource controllers.

My preference would be to do both. I’d like to be able to gate every PR on these tests, and that’s not realistic if we’re using real cloud provider APIs. I’m okay starting with just one (real APIs) or the other (mocks) to begin with, though.